src/main.ts

```diff
@@ -7,479 +7,1026 @@ import {
+      this.currentMatches = [];
+      return this.advance(keypress);
     }
 
-    this.leaderPending = false;
+    this.currentSequence.push(keypress);
+    const bestMatch = this.trie.bestMatch(this.currentSequence);
+    const matchKind = classifyMatch(bestMatch);
+    this.currentMatches = bestMatch ? bestMatch.leafValues() : [];
+
+    switch (matchKind) {
+      case MatchKind.NoMatch:
+        this.currentState = wasAlreadySearching
+          ? MatchState.InvalidMatch
+          : MatchState.EmptyMatch;
+        break;
+      case MatchKind.PartialMatch:
+        this.currentState = wasAlreadySearching
+          ? MatchState.ImprovedMatch
+          : MatchState.StartedMatch;
+        break;
+      case MatchKind.FullMatch:
+        this.currentState = wasAlreadySearching
+          ? MatchState.SuccessMatch
+          : // Very sus to reach success state at first try.
+            MatchState.SuccessMatch;
+        break;
+    }
+
+    return this.currentState;
+  };
+
+  public allMatches = (): readonly KeyMap[] => {
+    return this.currentMatches;
+  };
+
+  public fullMatch = (): Optional<KeyMap> => {
+    const numMatches = this.allMatches().length;
+    const isFullMatch = this.currentState === MatchState.SuccessMatch;
+
+    // Sanity checking.
+    if (isFullMatch && numMatches !== 1) {
+      writeConsole(
+        'State Machine in FullMatch state, but availableHotkeys.length contains more than 1 element. This is definitely a bug.',
+      );
+      return null;
+    }
+
+    if (isFullMatch && numMatches === 1) {
+      return this.currentMatches[0];
+    }
+    return null;
+  };
+
+  public stateKind = (): MatchStateKind => {
+    if (this.currentState === MatchState.EmptyMatch) {
+      return MatchStateKind.Initial;
+    }
+
+    const flowStates = [
+      MatchState.StartedMatch,
+      MatchState.RetainedMatch,
+      MatchState.ImprovedMatch,
+    ];
+
+    return flowStates.includes(this.currentState)
+      ? MatchStateKind.Flow
+      : MatchStateKind.Terminal;
   };
 }
 
-class SetHotkeyModal extends Modal {
-  private readonly currentLeader: string;
-  private readonly redraw: () => void;
-  private readonly setNewKey: (
-    key: string,
-    meta: boolean,
-    shift: boolean,
-  ) => void;
+class MatchHandler {
+  private trie: Trie<KeyMap>;
+  private machine: MatchMachine;
+  private readonly parent: LeaderHotkeys;
 
-  constructor(
-    app: App,
-    currentLeader: string,
-    redraw: () => void,
-    setNewKey: (newKey: string, meta: boolean, shift: boolean) => void,
-  ) {
-    super(app);
-    this.currentLeader = currentLeader;
-    this.redraw = redraw;
-    this.setNewKey = setNewKey;
+  public constructor(parent: LeaderHotkeys) {
+    this.parent = parent;
+    this.setKeymap(parent.settings.hotkeys);
   }
 
-  public onOpen = (): void => {
-    const { contentEl } = this;
+  public readonly handleKeyDown = (event: KeyboardEvent): void => {
+    console.log(event);
```

Comment:
We should probable remove this. Seems to log every key stroke in Obsidian right now.